mathics/builtin/lists.py

```diff
@@ -2510,6 +2518,8 @@ def __init__(self, df, p, evaluation):
     def _compute_distance(self, i, j):
         p = self._p
         d = apply_N(self._df(p[i], p[j]), self._evaluation)
```

Comment:
I see this pattern of apply_N followed by an if statement where the more general evaluate needs to be done many times in this PR. 

Why not just do this inside apply_N ?  Or have two kinds of apply_N ?  

More broadly it is this kind of activity that I feel down the line makes the code more complex because it is an idiom that you sort of have to figure out by looking at, and it is not clear to me that duplicating the code everywhere is that much of a win over just doing it once in a function somewhere.